Terminals.cs

```diff
@@ -0,0 +1,525 @@
+          if (string.CompareOrdinal(lexeme, "RETURN") == 0) {
+            return Token.RETURN;
+          } /* end if */
+          break;
+
+        case 'T' :
+          switch (lexeme[0]) {
+
+            case 'E' :
+              /* EXPORT */
+              if (string.CompareOrdinal(lexeme, "EXPORT") == 0) {
+                return Token.EXPORT;
+              } /* end if */
+              break;
+
+            case 'I' :
+              /* IMPORT */
+              if (string.CompareOrdinal(lexeme, "IMPORT") == 0) {
+                return Token.IMPORT;
+              } /* end if */
+              break;
+
+            case 'R' :
+              /* REPEAT */
+              if (string.CompareOrdinal(lexeme, "REPEAT") == 0) {
+                return Token.REPEAT;
+              } /* end if */
+              break;
+
+          } /* end switch */
+          break;
+
+      } /* end switch */
+      break;
+
+    case /* length = 7 */ 7 :
+      switch (lexeme[0]) {
+
+        case 'A' :
+          /* ARGLIST */
+          if (string.CompareOrdinal(lexeme, "ARGLIST") == 0) {
+            return Token.ARGLIST;
+          } /* end if */
+          break;
+
+        case 'P' :
+          /* POINTER */
+          if (string.CompareOrdinal(lexeme, "POINTER") == 0) {
+            return Token.POINTER;
+          } /* end if */
+          break;
+
+      } /* end switch */
+      break;
+
+    case /* length = 9 */ 9 :
+      switch (lexeme[0]) {
+
+        case 'P' :
+          /* PROCEDURE */
+          if (string.CompareOrdinal(lexeme, "PROCEDURE") == 0) {
+            return Token.PROCEDURE;
+          } /* end if */
+          break;
+
+        case 'Q' :
+          /* QUALIFIED */
+          if (string.CompareOrdinal(lexeme, "QUALIFIED") == 0) {
+            return Token.QUALIFIED;
+          } /* end if */
+          break;
+
+      } /* end switch */
+      break;
+
+    case /* length = 10 */ 10 :
+      /* DEFINITION */
+      if (string.CompareOrdinal(lexeme, "DEFINITION") == 0) {
+        return Token.DEFINITION;
+      } /* end if */
+      break;
+
+    case /* length = 14 */ 14 :
+      /* IMPLEMENTATION */
+      if (string.CompareOrdinal(lexeme, "IMPLEMENTATION") == 0) {
+        return Token.IMPLEMENTATION;
+      } /* end if */
+      break;
+
+  } /* end switch (length) */
+
+  return Token.Unknown;
+} /* end TokenForResword */
+
+
+/* ---------------------------------------------------------------------------
+ * method LexemeForResword(token)
+ * ---------------------------------------------------------------------------
+ * Returns a string with the lexeme for the reserved word represented by
+ * token.  Returns null if the token does not represent a reserved word.
+ * ------------------------------------------------------------------------ */
+
+public static string LexemeForResword (Token token) {
+  if (IsResword(token) == false) {
+    return null;
+  } /* end if */
+  return token.ToString();
+} /* end LexemeForResword */
```

Comment:
Although this matters less for reserved words, for we use interned strings for all lexemes within the compiler. For consistency we should also apply this to reserved words.

The reason for interning strings is that at any point within the compiler we can compare two strings with a single identity test instead of having to string-compare them.

For example, the identifiers of modules and functions in Modula-2 are repeated at the end of their declarations:

```modula-2
PROCEDURE Foobar ( baz : Bam );
...
END Foobar;
```

When checking that the END part matches the procedure header, we can simply test for identity:

```csharp
ident1 = Lexer.CurrentLexeme();
...
ident2 = LexerCurrentLexeme();
...
if (ident1 == ident2) {
  /* identifiers match */
}
```

If we use the `To.String()` method, the strings won't be interned.

In the C version, I defined private arrays that contain the strings, then calculate the index for the array, fetch the string and return it. Even though C# has methods for interning strings, it is preferable to use the same approach as in the C version. Ideally, the string arrays should be initialised at compile time.